context_propagation/interceptor.rb

```diff
@@ -0,0 +1,149 @@
+# frozen_string_literal: true
+
+require 'temporalio/client/interceptor'
+require 'temporalio/worker/interceptor'
+
+module ContextPropagation
+  class Interceptor
+    include Temporalio::Client::Interceptor
+    include Temporalio::Worker::Interceptor::Workflow
+    include Temporalio::Worker::Interceptor::Activity
+
+    def initialize(*keys_to_propagate)
+      @keys_to_propagate = keys_to_propagate
+    end
+
+    def intercept_client(next_interceptor)
+      ClientOutbound.new(self, next_interceptor)
+    end
+
+    def intercept_workflow(next_interceptor)
+      WorkflowInbound.new(self, next_interceptor)
+    end
+
+    def intercept_activity(next_interceptor)
+      ActivityInbound.new(self, next_interceptor)
+    end
+
+    def context_to_headers(input)
+      @keys_to_propagate.each do |key|
+        value = Thread.current[key]
+        input.headers[key.to_s] = value unless value.nil?
+      end
+    end
+
+    def with_context_from_headers(input)
+      # Grab all original values
+      orig_values = @keys_to_propagate.map { |key| [key, Thread.current[key]] }
+      # Replace values, even if they are nil
+      @keys_to_propagate.each { |key| Thread.current[key] = input.headers[key.to_s] }
+      begin
+        yield
+      ensure
+        # Put them all back, even if they were nil
+        orig_values.each { |key, val| Thread.current[key] = val }
```

Comment:
It would be setting a better example to create a single container within the thread-local namespace to hold the application-logic keys, rather than having them all at the the top-level.

Reply:
The user of the interceptor can do this (single key with a hash of all values needed), this interceptor is just a general purpose key copier. Would actually probably encourage users of this interceptor to not use multiple keys.

Reply:
I didn't understand your response here.